## [scSNViz](https://horvathlab.github.io/NGS/scSNViz) 
scSNViz plots information about scSNVs onto a dimensionally reduced representation.
## [scSNViz R package](https://github.com/HorvathLab/NGS/blob/8d57afcd3dfadf7c1ba3b7eb9e425763672b18ed/scSNViz%20R%20readme/index.md)
## [SCExecute](https://horvathlab.github.io/NGS/SCExecute) 
SCExecute generates cell-barcode specific BAM files from aligned, aggregate single-cell sequencing data, executing a user-provided command on each barcode-stratified BAM file.
## [SCReadCounts](https://horvathlab.github.io/NGS/SCReadCounts) 
SCReadCounts is a computational tool for a cell-level assessment of the read counts bearing a particular nucleotide at genomic positions of interest from single cell RNA sequencing (scRNA-seq) data. 
## [ReadCounts](https://horvathlab.github.io/NGS/ReadCounts) 
ReadCounts tabulates the number of reads providing evidence for variant and reference nucleotides at specific genomic loci and applies statistical tests to recognize allelic read-counts consistent with homozygous and heterozygous loci.
## [RNA2DNAlign](https://horvathlab.github.io/NGS/RNA2DNAlign)
RNA2DNAlign evaluates evidence for asymmetric allele distribution in next-gen sequencing reads of DNA and RNA samples from the same individual.
## [SNPlice](https://horvathlab.github.io/NGS/SNPlice)
SNPlice finds and evaluates the co-occurrence of single-nucleotide-polymorphisms (SNP) and altered splicing in next-gen mRNA sequence reads.
## [scDeseq2](https://github.com/HorvathLab/NGS/tree/master/scDeseq2)
scDeseq2 calculates differential gene expression for single cells using DESeq2.
