
ReadCounts tabulates the number of reads providing evidence for variant and reference nucleotides 
at specific genomic loci and applies statistical tests to recognize allelic 
read-counts consistent with homozygous and heterozygous loci.

ReadCounts is available as a self-contained binary package for 64-bit
Linux and MacOS/Darwin systems and as Python source. 
The pysam package, plus a variety
of common third-party python packages including numpy and scipy must
be installed to use ReadCounts in Python source form. See the install
instructions for more details. The self-contained binary package is
appropriate for most Linux and MacOS users.

Setup:
* [Download](https://github.com/HorvathLab/NGS/releases/ReadCounts-2.2.0)
* [Install](docs/Installation.md)

Usage:
* [ReadCounts](docs/Usage.md)

File Formats:
* [Input Files](docs/InputFiles.md)
* [Output Files](docs/OutputFiles.md)

[Examples](docs/Examples.md)
